module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,235 @@
+			return fmt.Errorf("could not add events to batch: %w", err)
+		}
+		p.log.Debug().Int("event_count", len(eventsList)).Msg("added events to batch")
+	}
+
+	return nil
+}
+
+// addResultsToBatch adds transaction results from in-memory storage to the batch.
+// If there are no results, this is a no-op.
+func (p *Persister) addResultsToBatch(batch storage.Batch) error {
+	if results := p.inMemoryResults.Data(); len(results) > 0 {
+		if err := p.results.BatchStore(p.executionResult.BlockID, results, batch); err != nil {
+			return fmt.Errorf("could not add transaction results to batch: %w", err)
+		}
+		p.log.Debug().Int("result_count", len(results)).Msg("added transaction results to batch")
+	}
+
+	return nil
+}
+
+// persistRegisters persists registers from in-memory to permanent storage.
+// If there are no registers, this is a no-op.
+func (p *Persister) persistRegisters() error {
+	if registerData := p.inMemoryRegisters.Data(); len(registerData) > 0 {
+		start := time.Now()
+
+		if err := p.registers.Store(registerData, p.header.Height); err != nil {
+			return fmt.Errorf("could not persist registers: %w", err)
+		}
+
+		p.log.Debug().
+			Int("register_count", len(registerData)).
+			Dur("duration_ms", time.Since(start)).
+			Msg("persisted registers")
+	}
+
+	return nil
+}
+
+// persistCollections persists collections from in-memory to permanent storage.
+// It skips collections that already exist in permanent storage.
+// If there are no collections, this is a no-op.
+func (p *Persister) persistCollections() error {
+	collections := p.inMemoryCollections.Collections()
+
+	if len(collections) == 0 {
+		return nil
+	}
+
+	start := time.Now()
+	persistedCount := 0
+
+	for _, collection := range collections {
+		light := collection.Light()
+		if err := p.collections.StoreLightAndIndexByTransaction(&light); err != nil {
+			if errors.Is(err, storage.ErrAlreadyExists) {
+				// Skip if already exists
+				p.log.Debug().
+					Hex("collection_id", logging.Entity(light)).
+					Msg("collection already exists in permanent storage")
+				continue
+			}
+			return fmt.Errorf("could not persist collection: %w", err)
+		}
+		persistedCount++
+	}
+
+	if persistedCount > 0 {
+		p.log.Debug().
+			Int("persisted_collections", persistedCount).
+			Int("total_collections", len(collections)).
+			Dur("duration_ms", time.Since(start)).
+			Msg("persisted collections")
+	}
+
+	return nil
+}
+
+// persistTransactions persists transactions from in-memory to permanent storage.
+// It skips transactions that already exist in permanent storage.
+// If there are no transactions, this is a no-op.
+func (p *Persister) persistTransactions() error {
+	transactions := p.inMemoryTransactions.Data()
+
+	if len(transactions) == 0 {
+		return nil
+	}
+
+	start := time.Now()
+	persistedCount := 0
+
+	for _, tx := range transactions {
+		if err := p.transactions.Store(&tx); err != nil {
+			if errors.Is(err, storage.ErrAlreadyExists) {
```

Comment:
is this also returned for transactions? the original logic only ran this for collections